### PR TITLE
P3io

### PR DESCRIPTION
**Pull request by chrispable**
Addresses issue #42 - This adds support to drive cabinet lights on a real ddr machine using REAL konami hardware provided you use libusb with it (the p3io konami io board). You can do this easily with zadig. Not only does it fully support of messages from the usb io device but the multiple serial light devices as well. Since different cabinets have different lights on different configurations and such, you will need to set options to tell the drivers what is where. Use the default settings to automatically setup a standard definition cabinet, the most common type in the wild. This also support the rarer deluxe cabinet with the light spires / satellites. Tested working in isolation. Will stress test at MGC 2017 but seems to be working fantastic (ableit full of hackery, someone please make this better lol). If merged, I'll supply my mods to tommath/tomcrypt (benefits everyone, eliminates 2 minute startup wait for entropy on windows with older runtimes, support for this made possible by a small tweak present in this code when working on portability for sm5, hence the prereq) and vs2010 project file (which is of dubious utility to others but works for me).

Highly suggest setting the following options in your ini if you use this on a ddr cab with windows, the following is what I use for my JP DDRX cabinet:

#ogl causes issues with 15/31 khz on a stock windows cab
VideoRenderers=d3d

#Lights are dumb as hell if you dont set these
BlinkGameplayButtonLightsOnNote=0
EasterEggs2=0

#This will help DDR players use the build
OnlyDedicatedMenuButtons=1

#Use only this so people can't put coins in with a keyboard, you cannot exit the game easily though!
#this also is an HDXB light driver. Due to the way this is architected, I cannot seperate HDXB into its own light device easily since NORMALLY it needs to talk to the p3io to work
InputDrivers=P3IO

#HD cabinet specific stuff, can also set SD or leave blank for SD
CabinetType=HD

#Give pad and SD neons as well as satellites
LightsDriver=EXTIO,Satellite

#set the proper com ports for the light services
EXTIOComPort=COM1
SatelliteComPort=COM2
COM4PORT=

#Note on COM4PORT, by default it uses a VIRTUAL com port on the p3io through bulk transfers.... yeah wtf.
#If you bypass this with a regular serial adapter, just put in something like COM4 as your COM4PORT. SHOULD work.
#If it doesn't, I added HDXB support to acrealio, you can make an HDXB device to stuff on a comport, fix it then put up a PR :)

### CODE REVIEW

**Comment by LightningXCE** [comment]
This sounds pretty great, thank you for your work on this.

**Comment by chrispable** [comment]
Forgot to mention all the HD code sports RGB light patterns which is why it's so messy, itg doesn't normally support RGB.

File: src/CryptHelpers.h [hunk 1]
```diff
@@ -1,6 +1,11 @@
 #ifndef CryptHelpers_H
 #define CryptHelpers_H
 
+//Workaround for new libtomcrypt variable names changing
+//this overrides the enum sadly, but I can't check for the enum in the preprocessor -- this is HIGHLY unlikely to ever change
+#define LTC_LTC_PKCS_1_V1_5 1
```
Comment (spigwitmer) [inline src/CryptHelpers.h]: * g++ (both 3.4 for arcade, and 6.3 for home) does not like this when it has been defined prior. sadly tomcrypt doesn't do us any favors keeping the same version number despite breaking compatibility. I'd do something like define-guard this against a custom preprocessor flag in `src/tomcrypt/src/header/tomcrypt_custom.h`, determining if it's our copy or the system's.
* CryptHelpers.cpp may be a better home for it as that's where tomcrypt.h is included anyway

File: src/Makefile.am [hunk 2]
```diff
@@ -146,7 +146,7 @@ LIBS += -lusb
 
 USBIO = io/USBDriver.cpp io/USBDriver.h io/PIUIO.cpp io/PIUIO.h io/ITGIO.cpp \
 	io/ITGIO.h io/PacDrive.cpp io/PacDrive.h io/G15.cpp io/G15.h \
-	io/USBDevice.h io/MiniMaid.cpp io/MiniMaid.h
+	io/USBDevice.h io/MiniMaid.cpp io/MiniMaid.h io/P3io.cpp io/P3IO.h
```
Comment (spigwitmer) [inline src/Makefile.am]: I think you mean `io/P3IO.cpp`

File: src/Makefile.am [hunk 3]
```diff
@@ -283,7 +287,16 @@ endif
 
 Arch = $(LoadingWindow) $(Sound) $(ArchHooks) $(InputHandler) $(MovieTexture) \
 	   $(Lights) $(MemoryCard) $(LowLevelWindow) $(ArchUtils) $(Dialog) $(Threads) $(USBDriver) \
-	   arch/arch_default.h arch/arch_platform.h arch/RageDriver.cpp arch/RageDriver.h
+	   arch/arch_default.h arch/arch_platform.h arch/RageDriver.cpp arch/RageDriver.h \
+	   arch/ACIO.h arch/ACIO.cpp arch/COM/serial.h arch/COM/serial.cc arch/COM/v8stdint.h
```
Comment (spigwitmer) [inline src/Makefile.am]: `arch/ACIO/ACIO.h arch/ACIO/ACIO.cpp` etc.etc.

File: src/Makefile.am [hunk 4]
```diff
@@ -283,7 +287,16 @@ endif
...
+
+if UNIX
+Arch += arch/COM/impl/unix.h arch/COM/impl/unix.cc arch/COM/impl/list_ports/list_ports_linux.cc
```
Comment (spigwitmer) [inline src/Makefile.am]: I had to add `-I$(srcdir)/arch/COM` to `AM_CXXFLAGS` (at the top near line 14) to get this to compile

File: src/arch/COM/impl/list_ports/list_ports_linux.cc [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+#include <cstdio>
+#include <cstdarg>
+#include <cstdlib>
+
+#include <glob.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "serial/serial.h"
```
Comment (spigwitmer) [inline src/arch/COM/impl/list_ports/list_ports_linux.cc]: had to change this to `#include "serial.h"` (alongside the AM_CXXFLAGS change) to get this to compile, though unsure if that'd break other build environments or external copies

File: src/io/P3IO.cpp [hunk 6]
```diff
@@ -0,0 +1,1315 @@
+#include "global.h"
+#include "Stepmania.h" //to assist with debugging
```
Comment (spigwitmer) [inline src/io/P3IO.cpp]: capital M in StepMania

File: src/arch/ACIO/ACIO.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+class ACIO
+{
+public:
+	static bool baudCheckWrapper(serial::Serial &acio_bus);
+	static bool baudCheck(serial::Serial &acio_bus);
+	static int prep_acio_packet_for_transmission(uint8_t* acio_request, int r_length, int init_escape_offset=1);
+	static int unescape_acio_packet(uint8_t* acio_response2, int len, int data_start=-1);
+	static int read_acio_packet(serial::Serial &acio_bus, uint8_t* acio_response);
+	static int write_acio_packet(serial::Serial &acio_bus, uint8_t* acio_request,int r_length);
+	static int ACIO::assemble_init_packet_to_write(uint8_t* acio_request, uint8_t* init_line);
```
Comment (spigwitmer) [inline src/arch/ACIO/ACIO.h]: is declaring this function as a member of ACIO while you're declaring ACIO itself necessary? gcc (6.3 at least) did not like that prefix, but I saw no special treatment and it compiled fine without it

File: src/CryptHelpers.h (hunk 1 above)
Comment (chrispable) [inline src/CryptHelpers.h]: That's fine, I have no idea how to implement this and thought it was a benign change. I'll omit this change but would like help rolling a solution. After this set of changes, I'll make another fork for tomcrypt stuff.

File: src/Makefile.am (hunk 3 above)
Comment (chrispable) [inline src/Makefile.am]: Yup, sorry 'lol windows'. It doesn't use this file and I was typing it out by hand. Oops. Thanks for catching it.

File: src/Makefile.am (hunk 4 above)
Comment (chrispable) [inline src/Makefile.am]: Not sure why if it's explicitly defined later on but I'll add it no problem.

File: src/arch/ACIO/ACIO.h (hunk 7 above)
Comment (chrispable) [inline src/arch/ACIO/ACIO.h]: This is a simply a copy and paste error on my part that MSVC didn't care about. Removed the prefix. There was indeed no need for it.

File: src/arch/COM/impl/list_ports/list_ports_linux.cc (hunk 5 above)
Comment (chrispable) [inline src/arch/COM/impl/list_ports/list_ports_linux.cc]: I'm doing similar for the OSX one. I didn't have access to those compilation platforms so I didn't see the flag.

File: src/io/P3IO.cpp (hunk 6 above)
Comment (chrispable) [inline src/io/P3IO.cpp]: lol windows case insensitivity. Fixed.

**Comment by chrispable** [comment]
Added the requested changes plus one more to fix macosx. Please note these files all have to be added to windows project files! The minimaid changes actually break windows compilation because they are not in the project files. It's easy to right click and add but we should probably work on that in a separate branch later.

**Comment by chrispable** [comment]
Just finished testing at MGC! Things went very well! Had some crashes from video codecs that itg didn't recognize from some packs, I'll investigate those but that was minor.

Ran into memory issues. Went from about 200 test songs to 2500. For those most part, the stock 1gb of RAM worked fine. I would still highly recommend upgrading it though.

Want to see the code in action? Enjoy before the copyright robots flag the music.
https://youtu.be/mBxoAGwtpSM

**Comment by spigwitmer** [comment]
lgtm, so merging. thanks again!

**Pull request merged**